src/crimson/common/gated.h

```diff
@@ -52,4 +53,38 @@ class Gated {
   seastar::gate pending_dispatch;
 };
 
-}// namespace crimson::common
+class gate_per_shard : public Gated {
+ public:
+  gate_per_shard() : gates(seastar::smp::count) {}
+  ~gate_per_shard() override = default;
+  gate_per_shard(gate_per_shard&&) = default;
+
+  template <typename Func, typename T>
+  inline void dispatch_in_background(const char* what, T& who, Func&& func) {
+    gates[seastar::this_shard_id()].dispatch_in_background(what, who, std::forward<Func>(func));
+  }
+
+  template <typename Func, typename T>
+  inline seastar::future<> dispatch(const char* what, T& who, Func&& func) {
+    return gates[seastar::this_shard_id()].dispatch(what, who, std::forward<Func>(func));
+  }
+
+  seastar::future<> close() override {
```

Comment:
Same here, probably we shouldn't allow to close a particular gate.

Reply:
done.